### PR TITLE
[jit] Respect order of Parameters in rnn.py

### PR DESCRIPTION
**Pull request by driazati**
Previously to get a list of parameters this code was just putting them in the reverse order in which they were defined, which is not always right. This PR allows parameter lists to define the order themselves. To do this parameter lists need to have a corresponding function that provides the names of the parameters.

### CODE REVIEW

File: torch/csrc/jit/script/init.cpp [hunk 1]
```diff
@@ -240,8 +240,8 @@ struct VISIBILITY_HIDDEN ConstantPythonTupleValue : public PythonValue {
 
 // Represents all the parameters of a module as a List[Tensor]
 struct VISIBILITY_HIDDEN ConstantParameterList : public SugaredValue {
-  ConstantParameterList(std::shared_ptr<Module> module)
-      : module_(std::move(module)) {}
+  ConstantParameterList(std::shared_ptr<Module> module, py::list params)
+      : module_(std::move(module)), params_(std::move(params)) {}
```
Comment (eellison) [inline torch/csrc/jit/script/init.cpp]: Maybe convert & store the  params to std::string here ? i imagine that's less overhead

File: torch/nn/modules/rnn.py [hunk 2]
```diff
@@ -134,9 +134,13 @@ def reset_parameters(self):
             init.uniform_(weight, -stdv, stdv)
 
     @_parameter_list
-    def get_flat_weights(self):
+    def _get_flat_weights(self):
         return self._flat_weights
 
+    def _get_flat_weights_names(self):
+        all_weights = [[weight for weight in weights] for weights in self._all_weights]
```
Comment (eellison) [inline torch/nn/modules/rnn.py]: imo the 4 weight(s) on one line is kidn of confusing

File: torch/nn/modules/rnn.py [hunk 3]
```diff
@@ -134,9 +134,13 @@ def reset_parameters(self):
...
+        return [p for layerparams in all_weights for p in layerparams]
```
Comment (eellison) [inline torch/nn/modules/rnn.py]: ditto with the two layerparams / ps

**Review by eellison — COMMENTED** [review]
Sorry what exactly was the issue previously and what is the new behavior?

File: torch/csrc/jit/script/init.cpp (hunk 1 above)
Comment (driazati) [reply torch/csrc/jit/script/init.cpp]: It'd be more since it would duplicate the strings, the `py::list` is just a thin wrapper around a Python list with the same reference semantics

**Comment by driazati** [comment]
@eellison clarified reasons in PR message

File: torch/nn/modules/rnn.py [hunk 4]
```diff
@@ -133,8 +133,11 @@ def reset_parameters(self):
         for weight in self.parameters():
             init.uniform_(weight, -stdv, stdv)
 
-    @_parameter_list
-    def get_flat_weights(self):
+    def _get_flat_weights_names(self):
+        return [weight for weights in self._all_weights for weight in weights]
```
Comment (eellison) [inline torch/nn/modules/rnn.py]: do you mean rewriting? This is hard to read

File: torch/_jit_internal.py [hunk 5]
```diff
@@ -165,13 +165,16 @@ def ignore(fn):
...
 
-def _parameter_list(fn):
+def _parameter_list(parameter_names_fn):
     """
     Decorator to denote that a function returns a list of all the parameters
     in a module
     """
-    fn._is_parameter_list = True
-    return fn
+    def decorator(fn):
```
Comment (eellison) [inline torch/_jit_internal.py]: Use functools wrap so that the docs & such remain n?

File: torch/csrc/jit/script/init.cpp [hunk 6]
```diff
@@ -322,9 +324,10 @@ struct ModuleValue : public SugaredValue {
     py::object py_module = py::cast(module);
     if (py::object attr = py::getattr(py_module, field.c_str(), py::none())) {
       if (py::isinstance<py::function>(attr) &&
-          py::hasattr(attr, "_is_parameter_list") &&
-          py::cast<bool>(py::getattr(attr, "_is_parameter_list"))) {
-        return std::make_shared<ConstantParameterList>(module);
+          py::hasattr(attr, "_parameter_names_fn")) {
```
Comment (eellison) [inline torch/csrc/jit/script/init.cpp]: Could you add a comment to this ?

File: test/test_jit.py [hunk 7]
```diff
@@ -10571,6 +10571,30 @@ def foo(x):
...
+                self.rnn = nn.LSTM(2, 3, 2)
+
+            def forward(self, x, lengths, h0, c0):
+                return self.rnn(x, (h0, c0))[0]
+
+        inputs = (torch.randn(1, 1, 2), torch.LongTensor([7]), torch.randn(2, 1, 3), torch.randn(2, 1, 3))
+        eager_out = self.runAndSaveRNG(lambda: Eager()(*inputs), ())[0]
+        script_out = self.runAndSaveRNG(lambda: M()(*inputs), ())[0]
+
+        self.assertEqual(eager_out, script_out)
```
Comment (eellison) [inline test/test_jit.py]: Maybe add a check that the export import run works as well ?

**Review by eellison — APPROVED** [review]
Looks good, a little hard to follow.

File: torch/_jit_internal.py (hunk 5 above)
Comment (driazati) [reply torch/_jit_internal.py]: It returns the function itself so those should all be preserved already

File: test/test_jit.py (hunk 7 above)
Comment (driazati) [reply test/test_jit.py]: The module hook should be thorough enough for this
